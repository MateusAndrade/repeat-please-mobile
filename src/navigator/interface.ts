import { IPAType } from '../../types/ipa';

export type RootStackParamList = {
	AuthenticatedNavigator: undefined;
	UnauthenticatedNavigator: undefined;
	Slides: undefined;
	Walkthrough: undefined;
	SignIn: undefined;
	SignUp: undefined;
	ForgotPassword: { email: string };
	Home: undefined;
	Levels: undefined;
	Challenge: { IPA: IPAType } | undefined;
	Result: undefined;
	Loading: undefined;
	TabNavigator: undefined;
};
