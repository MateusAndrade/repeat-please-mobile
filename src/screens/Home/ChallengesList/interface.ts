import { IPAType } from '../../../../types/ipa';
import { RootStackParamList } from '../../../navigator/interface';
import { StackNavigationProp } from '@react-navigation/stack';

export interface IChallengeListProps {
	navigation: StackNavigationProp<RootStackParamList, 'Levels'>;
	wordsToPratice: IPAType[];
	wordsToReview: IPAType[];
	bestResults: IPAType[];
}
