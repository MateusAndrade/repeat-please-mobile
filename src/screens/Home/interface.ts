import { NavigationStackProp } from 'react-navigation-stack';

export interface ILevelsProps {
	navigation: NavigationStackProp;
}
