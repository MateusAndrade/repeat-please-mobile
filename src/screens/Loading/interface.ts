import { RootStackParamList } from '../../navigator/interface';
import { StackScreenProps } from '@react-navigation/stack';

export type ILoadingProps = StackScreenProps<RootStackParamList, 'Loading'>;
