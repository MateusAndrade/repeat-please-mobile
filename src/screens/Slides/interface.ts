import { RootStackParamList } from '../../navigator/interface';
import { StackScreenProps } from '@react-navigation/stack';

export type ISlidesProps = StackScreenProps<RootStackParamList, 'Slides'>;
