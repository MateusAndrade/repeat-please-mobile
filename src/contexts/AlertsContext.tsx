import React, { createContext, useState, useContext } from 'react';
import { Alerts } from 'repeat-please-styles';
import {
	IAlertsState,
	IALertsContext,
	IProviderProps,
	IAlertsProps,
} from './interface';
import { ALERT_TIME_OPENED } from '../constants/alerts';

const initialState: IAlertsState = {
	shown: false,
	message: '',
	type: 'error',
};

export const AlertsContext = createContext<IALertsContext>({
	showAlert: () => undefined,
});

export const AlertsProvider = ({ children }: IProviderProps) => {
	const [state, setState] = useState(initialState);

	const hideAlert = (props: IAlertsProps = state) => {
		setState({ ...props, shown: false });
	};

	const showAlert = (props: IAlertsProps) => {
		setState({ ...props, shown: true });
		const timeoutId = setTimeout(() => {
			hideAlert(props);
			clearTimeout(timeoutId);
		}, ALERT_TIME_OPENED);
	};

	return (
		<AlertsContext.Provider value={{ showAlert }}>
			{children}
			<Alerts {...{ ...state, onCloseModal: hideAlert }} />
		</AlertsContext.Provider>
	);
};

export const useAlertContext = () => useContext(AlertsContext);
