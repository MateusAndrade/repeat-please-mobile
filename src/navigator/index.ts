import { createAppContainer, createSwitchNavigator } from 'react-navigation';
import { createStackNavigator } from 'react-navigation-stack';
import Slides from '../screens/slides';
import Walkthrough from '../screens/walkthrough';
import Login from '../screens/login';
import Register from '../screens/register';
import Home from '../screens/home';
import Loading from '../screens/loading';

const navigationOptions = () => ({ header: null });

const AppStack = createStackNavigator({
	Home: {
		screen: Home,
		navigationOptions,
	},
});

const AuthStack = createStackNavigator(
	{
		Slides: {
			screen: Slides,
			navigationOptions,
		},
		Walkthrough: {
			screen: Walkthrough,
			navigationOptions,
		},
		Login: {
			screen: Login,
		},
		Register: {
			screen: Register,
		},
	},
);

const navigator = createSwitchNavigator(
	{
		Loading,
		App: AppStack,
		Auth: AuthStack,
	},
	{
		initialRouteName: "Loading"
	},
);

export default createAppContainer(navigator);
