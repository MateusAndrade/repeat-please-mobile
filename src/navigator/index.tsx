import React from 'react';
import { createStackNavigator } from '@react-navigation/stack';
import { Loading } from '../screens/Loading';
import { TabNavigator } from './AuthenticatedNavigator/TabNavigator';
import { UnauthenticatedNavigator } from './UnauthenticatedNavigator';
import { NavigationContainer } from '@react-navigation/native';
import * as ROUTES from './routes';

const Stack = createStackNavigator();

export const AppNavigationContainer = () => (
	<NavigationContainer>
		<Stack.Navigator
			initialRouteName={ROUTES.LOADING}
			screenOptions={{ headerShown: false }}>
			<Stack.Screen name={ROUTES.LOADING} component={Loading} />
			<Stack.Screen name={ROUTES.TAB_NAVIGATOR} component={TabNavigator} />
			<Stack.Screen
				name={ROUTES.UNAUTHENTICATED}
				component={UnauthenticatedNavigator}
			/>
		</Stack.Navigator>
	</NavigationContainer>
);
