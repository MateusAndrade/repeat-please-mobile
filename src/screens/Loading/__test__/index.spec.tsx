import React from 'react';
import { render } from '@testing-library/react-native';
import { Loading } from '../';
import * as redux from 'react-redux';
import { ILoadingProps } from '../interface';
import { TAB_NAVIGATOR, UNAUTHENTICATED } from '../../../navigator/routes';
import * as AlertsContext from '../../../contexts/AlertsContext';

const defaultProps: any = {
	navigation: {
		navigate: jest.fn(),
	},
};

const setup = (props: ILoadingProps) => render(<Loading {...props} />);

describe('#Loading', () => {
	afterEach(() => {
		jest.clearAllMocks();
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	describe('#Unit Test', () => {
		it('renders the component properly', () => {
			jest.spyOn(redux, 'useSelector').mockReturnValue({
				profile: null,
			});
			jest.spyOn(redux, 'useDispatch').mockReturnValue(jest.fn());
			const wrapper = setup(defaultProps);
			expect(wrapper.toJSON()).toMatchSnapshot();
		});

		describe('when the profile is null', () => {
			it('dispatchs a function to get the profile', () => {
				jest.spyOn(redux, 'useSelector').mockReturnValue({
					profile: null,
				});
				const spy = jest.spyOn(redux, 'useDispatch').mockReturnValue(jest.fn());
				setup(defaultProps);
				expect(spy).toHaveBeenCalled();
			});
		});

		describe('when exists a profile', () => {
			[
				{ isFirstAccess: true, route: UNAUTHENTICATED },
				{ isFirstAccess: false, route: TAB_NAVIGATOR },
			].forEach(({ isFirstAccess, route }) => {
				describe(`when ${isFirstAccess ? 'is' : 'is not'} first access`, () => {
					it(`navigate to the ${route} route `, () => {
						jest.spyOn(redux, 'useSelector').mockReturnValue({
							profile: { isFirstAccess },
						});
						jest.spyOn(redux, 'useDispatch');

						setup(defaultProps);

						expect(defaultProps.navigation.navigate).toHaveBeenCalledWith(
							route,
						);
					});
				});
			});
		});

		describe('when occours an error to load the profile', () => {
			it('shows an error alert', () => {
				const showAlertMock = jest.fn();
				jest.spyOn(AlertsContext, 'useAlertContext').mockImplementation(() => ({
					showAlert: showAlertMock,
				}));
				jest.spyOn(redux, 'useSelector').mockReturnValue({
					profile: null,
					error: 'Error to Load the user profile',
				});
				jest.spyOn(redux, 'useDispatch');

				setup(defaultProps);

				expect(showAlertMock).toHaveBeenCalledWith({
					type: 'error',
					message: 'errorLoading',
				});
			});
		});
	});
});
