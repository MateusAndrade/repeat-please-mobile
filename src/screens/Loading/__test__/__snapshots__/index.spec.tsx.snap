// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Loading #Unit Test renders the component properly 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
